### ISSUE TITLE
Memory leak when the producer uses Headers

### ISSUE THREAD

**Issue by lintang0** [confluentinc/confluent-kafka-go#183]
Description
===========
When the producer sends kafka message with "Headers", the memory usage keeps going up. No memory leak if no Headers inside the message.

How to reproduce
================
run the following code, which is modified from the sample producer code, memory leak can be reprodued.

    package main

    /**
      Memory leak detected when Headers inside message
    */

    import (
	"fmt"
	"os"

	"github.com/confluentinc/confluent-kafka-go/kafka"
    )

    func main() {

	if len(os.Args) != 3 {
		fmt.Fprintf(os.Stderr, "Usage: %s <broker> <topic>\n",
			os.Args[0])
		os.Exit(1)
	}

	broker := os.Args[1]
	topic := os.Args[2]

	p, err := kafka.NewProducer(&kafka.ConfigMap{"bootstrap.servers": broker})

	if err != nil {
		fmt.Printf("Failed to create producer: %s\n", err)
		os.Exit(1)
	}

	fmt.Printf("Created Producer %v\n", p)
	value := "Hello Go!"

	for {
		err = p.Produce(&kafka.Message{
			TopicPartition: kafka.TopicPartition{Topic: &topic, Partition: kafka.PartitionAny},
			Value:          []byte(value),
			Headers:        []kafka.Header{{"myTestHeader", []byte("header values are binary")}},
		}, nil)
	}
}
Checklist
=========
Please provide the following information:

 - [x] confluent-kafka-go and librdkafka version (`LibraryVersion()`): 0.11.4
 - [x] Apache Kafka broker version: 2.11 confluent-3.2.2
 - [ ] Client configuration: `ConfigMap{...}`
 - [x] Operating system: Ubuntu 14.04
 - [ ] Provide client logs (with `"debug": ".."` as necessary)
 - [ ] Provide broker log excerpts
 - [x] Critical issue

**Comment by billygout** [comment]
Hi @edenhill. I'm working with @lintang0 here. I just wanted to check that the information here is sufficient for debugging purposes. Thanks!

**Comment by mhowlett** [comment]
makes me wonder - have we ever tried testing librdkafka with valgrind?

**Comment by billygout** [comment]
@mhowlett I see valgrind mentioned here: https://github.com/edenhill/librdkafka/tree/master/tests
Some allocations are done as cgo code in confluent-kafka-go, which is external to librdkafka, so even assuming librdkafka does get valgrind treatment in testing, it's possible to have leaks from that code. Not sure whether that's what is going on here...

**Comment by edenhill** [comment]
@mhowlett librdkafka is valgrind, helgrind and asan clean. I'm guessing this memory leak is in the interaction with the Go client.

**Comment by edenhill** [comment]
@billygout Will look into this first thing next week, thanks for a detailed bug report!

**Comment by billygout** [comment]
@edenhill Thanks, So far I've added printf's to `rd_kafka_headers_new` and `rd_kafka_headers_destroy`, and confirmed that those always come in corresponding pairs, so the librdkafka internal representation of the headers seem to be getting allocated and freed properly (assuming `rd_list_destroy(&hdrs->rkhdrs_list)` does the deep free i'd expect it to do, right?).

One thing I noticed is that while *both* the header key and header value of the `tmphdrs` are malloc'd here https://github.com/confluentinc/confluent-kafka-go/blob/master/kafka/producer.go#L221-L228 is malloc'd here (according to this doc https://golang.org/cmd/cgo/ which says that both `C.CBytes` and `C.CString` uses malloc and must be manually freed), I only see that the value is being de-allocated, and the de-allocation of the key seems to be missing here: https://github.com/confluentinc/confluent-kafka-go/blob/master/kafka/producer.go#L52. But even when I add `free((void *)tmphdrs[i].key);` there, I still see memory rising, so, perhaps that was one leak and there's yet another one?
